### PR TITLE
feat: multi-hop step accessor safety

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Follows-up on https://github.com/shapeshift/web/pull/6765 with `getHopByIndex` introducing hop access logic consolidation and safety:

- Adds a new `SupportedTradeQuoteStepIndex` so we make accessors safe by only being to access items at 0th and 1st position, since we don't support multi-hops of more than 2 hops currently). This only ensures easier migration to multi-hops consisting of more than 2 hops when we do support that.
- Adds a new `getHopByIndex` that:
  - narrows the return type by the input `quote` being undefined (or not)
  - leverages the type-safety above to ensure we only access elements that aren't out of bounds
  - ensures we don't try to access an index > 0 for single-hop quotes
 

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- relates to the closed https://github.com/shapeshift/web/issues/6764

> What protocols, transaction types or contract interactions might be affected by this PR?

Very low, this should come as no runtime change, do a small regression of swapper end-to-end as paranoia

## Testing

- Swapper is still happy for single hops
- Swapper is still happy for Li.Fi multi-hops (see PR under this one i.e https://github.com/shapeshift/web/pull/6765)

### Engineering

- ☝🏽 

### Operations

- ☝🏽

### CODE REVIEW

File: src/state/slices/tradeQuoteSlice/helpers.ts [hunk 1]
```diff
@@ -106,6 +117,56 @@ export const getTotalProtocolFeeByAsset = (quote: TradeQuote): Record<AssetId, P
     {},
   )
 
-export const sortQuotes = (unorderedQuotes: ApiQuote[]): ApiQuote[] => {
+const sortApiQuotes = (unorderedQuotes: ApiQuote[]): ApiQuote[] => {
   return orderBy(unorderedQuotes, ['inputOutputRatio', 'swapperName'], ['desc', 'asc'])
 }
+
+export const sortTradeQuotes = (
+  tradeQuotes: Partial<Record<SwapperName, Record<string, ApiQuote>>>,
```
Comment (0xApotheosis) [inline src/state/slices/tradeQuoteSlice/helpers.ts:125]:
```suggestion
  tradeQuotes: PartialRecord<SwapperName, Record<string, ApiQuote>>,
```

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
I'm currently not able to get quotes on this branch, tried both UTXOs and EVMs, with single and cross-account trades.

Screenshot 2024-05-02 at 3 18 23 PM

Screenshot 2024-05-02 at 3 20 18 PM

File: src/state/slices/tradeQuoteSlice/helpers.ts (hunk 1 above)
Comment (gomesalexandre) [reply src/state/slices/tradeQuoteSlice/helpers.ts]: 0f5a515299

**Comment by gomesalexandre** [comment]
> I'm currently not able to get quotes on this branch, tried both UTXOs and EVMs, with single and cross-account trades.

@0xApotheosis as discussed, way too much "safety" in this PR indeed, voiding very valid accessors at runtime. Fixed in 74ad060060 - and also removed the whole generics dance while at it. The whole `if trade quote is defined, hop is defined` was actually invalid. 

This prompted us to "fix" types in c437ee2da686cdb40220e4ec464f2dd3093eee20 - note `fix` in quotation marks here as most of this is really explicitly checking for undefineds - or adding non-null assertion for first and last hops - which we were not doing before.

Confirmed single hop quotes are happy again: 
image

And so are multi-hops:

image


Also did a full e2e regression test of a single-hop:

https://github.com/shapeshift/web/assets/17035424/27b9faa3-cd84-4b5a-9c89-94ba1ad15550

And of a multi-hop:

image

**Review by 0xApotheosis — APPROVED** [review]
There we go ✅ 

Checked both single and multi-hop trades on both EVM and UTXO assets.
All are now happy.

Screenshot 2024-05-03 at 9 58 05 AM

**Pull request merged**